### ISSUE TITLE
[onert] Reducing peak memory usage during model loading

### ISSUE THREAD

**Issue by periannath** [Samsung/ONE#2077]
onert consumes almost twice as much memory as the model size during model loading.

### Model loading process
1. Read a whole model file into memory (`_buffer` variable)
1. Load operators from `_buffer` and create Graph
1. Load weight and bias data and copy them into `CachedData` :point_left: **Peak memory**
1. Deallocate `_buffer` from memory

At step 3, weight and bias data are duplicated at `_buffer` and `CachedData`.

### How to reduce peak memory usage? (list of options)

1. Deallocate memory area for weight and bias data when it is copied into `CachedData`
   - Later phases are not affected
   - How to deallocate memory area for weight and bias data?
1. Using `mmap` to read model file and `CachedData` has `mmap` address
   - Deallocation using `munmap` is needed at dtor of `CachedData`
   - Any problem when nnapi is used?

**Comment by periannath** [comment]
### Profiling memory usage for option 1

- Commit : #2097 
- Model : inception_v3_2018_04_27
- command
	```bash
	$ BACKENDS=cpu valgrind --pages-as-heap=yes --tool=massif --detailed-freq=1 ./Product/armv7l-linux.debug/out/bin/nnpackage_run --nnpackage ../nnpkg_tst/inception_v3_2018_04_27/
	```

Using valgrind to profile memory usage for `nnpackage_run`
Valgrind will terminate after model prepare because some neon instruction is not supported at valgrind

**onert master**

![mass_no_mmap_point](https://user-images.githubusercontent.com/10445076/84240913-ab153200-ab39-11ea-9896-968cdd8d61db.png)

**onert draft**

![massif_mmap_point](https://user-images.githubusercontent.com/10445076/84240895-a6e91480-ab39-11ea-913d-ce926f574096.png)

- Memory usage during model loading is almost same as model size

**Comment by periannath** [comment]
### Measure model load time for option 1

**onert master**
```bash
$ BACKENDS=cpu ./Product/armv7l-linux.release/out/bin/nnpackage_run --nnpackage ../nnpkg_tst/inception_v3_2018_04_27
... run 0 takes 1263.64 ms
===================================
MODEL_LOAD takes 411.48 ms
PREPARE    takes 281.398 ms
EXECUTE    takes 
- Min:  1263.64 ms
- Max:  1263.64 ms
- Mean: 1263.64 ms
- GeoMean: 1263.65 ms
===================================
```


**onert draft**
```bash
$ BACKENDS=cpu ./Product/armv7l-linux.release/out/bin/nnpackage_run --nnpackage ../nnpkg_tst/inception_v3_2018_04_27
... run 0 takes 1249.76 ms
===================================
MODEL_LOAD takes 248.499 ms
PREPARE    takes 232.491 ms
EXECUTE    takes 
- Min:  1249.76 ms
- Max:  1249.76 ms
- Mean: 1249.76 ms
- GeoMean: 1249.76 ms
===================================
```

- Model load time is reduced
  - Master : 411 ms
  - Draft : 248 ms
  - Additional memory allocation for `_buffer` is removed

**Comment by periannath** [comment]
### How to implement option 2
- `ExternalData` has `base` pointer ans its size

#### Implementation 
1. mmap model file (option 1)
1. Replacing `CachedData` into `ExternalData`
   - `ExternalData` has pointer to mmaped address
1. Deallocation using `munmap` is not yet implemented

**Comment by periannath** [comment]
### Model load time and memory usage (option 2)

- Commit: [836f297](https://github.com/Samsung/ONE/pull/2097/commits/836f29734d9fc70c0863ae581bb02dfee7c139f2)

| Runtime | Load time memory | Load time | Prepare time |
|--|--|--|--|
| master | 187 MB | 419 ms | 288 ms |
| option 1 | 100 MB | 243 ms | 254 ms |
| option 2 | 14 MB | 6 ms | 290 ms |

- Peak memory usage for load time is reduced
- Model load time is reduced
  - Additional memory allocation for `_buffer` is removed

<details><summary>Raw data from nnpackage_run</summary>

**onert master**
```bash
$ BACKENDS=cpu ./Product/armv7l-linux.release/out/bin/nnpackage_run --nnpackage ../nnpkg_tst/inception_v3_2018_04_27 --mem_poll=true
... run 0 takes 1283.27 ms
===================================
MODEL_LOAD takes 419.034 ms
PREPARE    takes 288.84 ms
EXECUTE    takes
- Min:  1283.27 ms
- Max:  1283.27 ms
- Mean: 1283.27 ms
- GeoMean: 1283.27 ms
===================================
RSS
- MODEL_LOAD takes 191704 kb
- PREPARE    takes 203244 kb
- EXECUTE    takes 134940 kb
- PEAK       takes 203244 kb
===================================
HWM
- MODEL_LOAD takes 194080 kb
- PREPARE    takes 203808 kb
- EXECUTE    takes 203808 kb
- PEAK       takes 203808 kb
===================================
```

**option 1**
```bash
$ BACKENDS=cpu ./Product/armv7l-linux.release/out/bin/nnpackage_run --nnpackage ../nnpkg_tst/inception_v3_2018_04_27 --mem_poll=true
... run 0 takes 1252.29 ms
===================================
MODEL_LOAD takes 253.247 ms
PREPARE    takes 254.015 ms
EXECUTE    takes 
- Min:  1252.29 ms
- Max:  1252.29 ms
- Mean: 1252.29 ms
- GeoMean: 1252.29 ms
===================================
RSS
- MODEL_LOAD takes 101928 kb
- PREPARE    takes 203964 kb
- EXECUTE    takes 137896 kb
- PEAK       takes 203964 kb
===================================
HWM
- MODEL_LOAD takes 102936 kb
- PREPARE    takes 204468 kb
- EXECUTE    takes 204468 kb
- PEAK       takes 204468 kb
===================================
```

**option 2**
```bash
$ BACKENDS=cpu ./Product/armv7l-linux.release/out/bin/nnpackage_run --nnpackage ../nnpkg_tst/inception_v3_2018_04_27 -m1
... run 0 takes 1235.91 ms
===================================
MODEL_LOAD takes 6.275 ms
PREPARE    takes 295.026 ms
EXECUTE    takes
- Min:  1235.91 ms
- Max:  1235.91 ms
- Mean: 1235.91 ms
- GeoMean: 1235.91 ms
===================================
RSS
- MODEL_LOAD takes 13984 kb
- PREPARE    takes 200916 kb
- EXECUTE    takes 229052 kb
- PEAK       takes 229052 kb
===================================
HWM
- MODEL_LOAD takes 13984 kb
- PREPARE    takes 200916 kb
- EXECUTE    takes 229340 kb
- PEAK       takes 229340 kb
===================================
```

</details>

**Comment by YongseopKim** [comment]
Personally I've tried this further for using ExternalData all phases(ModelLoad~Execute). https://github.com/YongseopKim/ONE/commit/e69aa3a0c674fb8c74f449175b352329e299f280

## Inv3

Memory

| version | ModelLoad | Prepare | Execute |
|---------|-----------|---------|---------|
| master  | 192 MB    | 203 MB  | 135 MB  |
| draft   | 10 MB     | 21 MB   | 204 MB  |

Latency

| version | ModelLoad | Prepare | Execute |
|---------|-----------|---------|---------|
| master  | 400 ms    | 229 ms  | 1271 ms |
| draft   | 71 ms     | 109 ms  | 1310 ms |

## Mv2

Memory

| version | ModelLoad | Prepare | Execute |
|---------|-----------|---------|---------|
| master  | 345 MB    | 413 MB  | 344 MB  |
| draft   | 7 MB      | 13 MB   | 421 MB  |

Latency

| version | ModelLoad | Prepare | Execute |
|---------|-----------|---------|---------|
| master  | 51 ms     | 42 ms   | 277 ms  |
| draft   | 4 ms      | 14 ms   | 286 ms  |

## other models

The draft can't run the other models for now.

**Comment by YongseopKim** [comment]
The above table seems too weird... I'll update it soon.

**Comment by YongseopKim** [comment]
I've profiled my draft version because I thought the peak of rss is too high(223mb).

tl;dr
- I can't answer the reason why the peak of rss is too high.
- However I guess that not main threads copy the mapped memory from main thread.
- If a goal is to lower times of model laod and prepare, this is a good option.
- If a goal is to lower memory of overall, it's better to focus on lowering memory of prepare phase by mixed external data & cached data & unmap.

Default thread number
- tflite: 4
- onert: 8

Latency
- unit: ms
- warming up 5 & running 5

|                 | MODEL_LOAD | PREPARE | EXECUTE |
|-----------------|------------|---------|---------|
| tflite          | 16         | 0.5     | 1565    |
| onert-master    | 235        | 1000    | 1278    |
| onert-draft-T=8 | 33         | 822     | 1248    |
| onert-draft-T=4 | 33         | 784     | 1331    |
| onert-draft-T=2 | 35         | 810     | 1494    |
| onert-draft-T=1 | 32         | 814     | 2155    |

RSS
- unit: kb
- physically on ram

|                 | MODEL_LOAD | PREPARE | EXECUTE |
|-----------------|------------|---------|---------|
| tflite          | 8580       | 9752    | 228017  |
| onert-master    | 101760     | 204152  | 136860  |
| onert-draft-T=8 | 7132       | 22236   | 223887  |
| onert-draft-T=4 | 7128       | 21440   | 223615  |
| onert-draft-T=2 | 7188       | 22756   | 224579  |
| onert-draft-T=1 | 7128       | 22232   | 205872  |

VMS
- unit: kb
- logically on ram

|                 | MODEL_LOAD | PREPARE | EXECUTE |
|-----------------|------------|---------|---------|
| tflite          | 173016     | 305596  | 352232  |
| onert-master    | 121924     | 217348  | 242270  |
| onert-draft-T=8 | 113740     | 123228  | 325640  |
| onert-draft-T=4 | 113868     | 123224  | 288756  |
| onert-draft-T=2 | 113736     | 123224  | 270276  |
| onert-draft-T=1 | 113736     | 123224  | 226891  |

If you're interested in, please see https://github.com/YongseopKim/ONE/issues/2

**Comment by periannath** [comment]
### Investigate peak memory usage of above draft

Draft of @YongseopKim : https://github.com/YongseopKim/ONE/tree/test/use_external_data_pulled

I've profiled above draft using valgrind on x86_64 system. Valgrind on x86_64 runs well without any error.

#### Memory usage of master

![inception_master](https://user-images.githubusercontent.com/10445076/85371024-9b024700-b56a-11ea-9bec-a8f6a5e67d83.png)

#### Memory usage of draft

![inception_draft](https://user-images.githubusercontent.com/10445076/85371032-9e95ce00-b56a-11ea-84a0-4f26888de521.png)

#### Cause of memory peak in inference

- Convolution layer allocates additional memory to copy `filter_data`
- Master : Copied `filter_data` in `CachedData` is deallocated
- Draft : Copied `filter_data` in `ExternalData` is not unmapped

#### Solution (?)

- Deallocate `ExternalData` after it is copied for convolution

[massif.inception.zip](https://github.com/Samsung/ONE/files/4817620/massif.inception.zip)

**Comment by YongseopKim** [comment]
Accroding to @periannath 's advice, I can remove the external data on the execute phase.

draft2: https://github.com/YongseopKim/ONE/tree/test/use_external_data_pulled

Latency
- unit: ms
- warming up 5 & running 5

|                 | MODEL_LOAD | PREPARE | EXECUTE |
|-----------------|------------|---------|---------|
| tflite          | 16         | 0.5     | 1565    |
| onert-master    | 235        | 1000    | 1278    |
| onert-draft     | 33         | 822     | 1248    |
| onert-draft2    | 33         | 769     | 1251    |

RSS
- unit: kb
- physically on ram

|                 | MODEL_LOAD | PREPARE | EXECUTE |
|-----------------|------------|---------|---------|
| tflite          | 8580       | 9752    | 228017  |
| onert-master    | 101760     | 204152  | 136860  |
| onert-draft     | 7132       | 22236   | 223887  |
| onert-draft2    | 7292       | 21860   | 132173  |

**Comment by periannath** [comment]
All related PRs merged. Please see #2580 if someone is interested in using `ExternalData` instead of `CachedData`.
